pre_commit/languages/download.py

```diff
@@ -0,0 +1,258 @@
+
+
+@dataclass(frozen=True)
+class Platform:
+    value: str
+
+    def __post_init__(self) -> None:
+        os, cpu = self.parts
+
+        if os not in {'linux', 'darwin', 'windows'}:
+            raise ValueError(f'invalid operating system `{os}`')
+
+        if cpu not in {'amd64', 'aarch64'}:
+            raise ValueError(f'invalid CPU `{cpu}`')
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> Literal[
+        'linux', 'darwin', 'windows', 'dragonfly',
+        'freebsd',
+    ]:
+        os, _ = self.parts
+        return os  # type: ignore
+
+    @property
+    def cpu(self) -> Literal[
+        'arm64', 'arm64be', 'arm', '386', 'amd64', 'ppc',
+        'ppc64', 'ppc64le',
+    ]:
+        _, cpu = self.parts
+        return cpu  # type: ignore
+
+    @staticmethod
+    def host() -> Platform:
+        os = {
+            'Linux': 'linux',
+            'Darwin': 'darwin',
+            'Windows': 'windows',
+            'DragonFly': 'dragonfly',
+            'FreeBSD': 'freebsd',
+        }[platform.system()]
+        cpu = {
+            'aarch64': 'arm64',
+            'aarch64_be': 'arm64be',
+            'arm': 'arm',
+            'i386': '386',
+            'i686': '386',
+            'x86_64': 'amd64',
+            'ppc': 'ppc',
+            'ppc64': 'ppc64',
+            'ppc64le': 'ppc64le',
+        }[platform.machine()]
+        return Platform(f'{os}/{cpu}')
+
+
+class ContentCorruptionError(Exception):
+    pass
+
+
+class SRI:
+    def __init__(self, value: str):
+        self._value = value
+        self._algorithm, self._checksum = self.value.split('-', 1)
+        assert self.algorithm in {'sha256', 'sha384', 'sha512'}
+
+        if b64encode(b64decode(self.checksum)).decode('utf-8') !=\
+                self.checksum:
+            raise ValueError('Invalid checksum string.')
+        if (self.algorithm == 'sha256' and len(self.checksum) != 44) or \
+           (self.algorithm == 'sha384' and len(self.checksum) != 64) or \
+           (self.algorithm == 'sha384' and len(self.checksum) != 88):
+            raise ValueError(
+                f'Invalid checksum string length for {self.algorithm}',
+            )
+
+        self._hasher = hashlib.new(self.algorithm)
+
+    @property
+    def value(self) -> str:
+        return self._value
+
+    @property
+    def algorithm(self) -> str:
+        return self._algorithm
+
+    @property
+    def checksum(self) -> str:
+        return self._checksum
+
+    @property
+    def digest(self) -> str:
+        return b64encode(self._hasher.digest()).decode('utf-8')
+
+    def update(self, data: bytes) -> None:
+        return self._hasher.update(data)
+
+    def check_content(self) -> None:
+        if self.digest != self.checksum:
+            raise ContentCorruptionError(f"""SRI checker content corruption,
+the digest is {self.algorithm}-{self.digest}""")
```

Comment:
The separation of concerns do not scale well here because you want one static SRI value and then validate multiple files.

I would suggest having SRI class a static representation of the subresource integrity rather than trying to do the hashing as well. The download function below can do the hashing part, it does not need to be part of the class. 

I would recommend changing this to just be accessors that mirror the `Hash` object accessors, i.e. it can be used an immutable `Hash` protocol:

```suggestion
    def digest(self) -> bytes:
        return b64_decode(self.checksum)

    def hexdigest(self) -> str:
        return self.digest.hex()
```

The hashing part can be done within a free function, or other helper functions.

Reply:
> you want one static SRI value and then validate multiple files

What do you mean? I thought one SRI value is unique to one file.

Reply:
The SRI is not re-entrant:

```py
sri = SRI(...)
with open(...) as stream:
    sri.update(stream.read())
sri.check_content()
with open(...) as stream:
    sri.update(stream.read())  # cannot re-use or reset hasher state.
sri.check_content()
```

It _could_ be best to represent this as a single _validate_ function:

```py
sri = SRI(...)
sri.validate(stream)
sri.validate(stream)
```

However, that removes the ability to do piecemeal update of the hasher with chunks. That can be done by constructing a hasher from the algorithm but then you'd need a API for constructing the hasher. You would then need a "check" function, or similar, to validate and raise the exception. The scope of the class then keeps increasing.

It _may_ be best to just keep the SRI class focussed on *one thing*: representing a static SRI string value. Move other things into either immutable member functions or free functions. It is always easier when a class has a tightly constrained scope, as it is easy to stop it expanding to a huge amount of functionality.

Reply:
AFAICT, ``hashlib`` objects are not meant to be re-entrant. In fact, there is even ``hash.copy()`` which returns a clone of the hash object. The intended use is to efficiently compute the digests of data sharing a common initial substring. 

A quick search on the Internet seems to suggest that you create a new hashlib object if you want to hash a new file: 
https://stackoverflow.com/questions/50535925/reusing-hashlib-md5-calculates-different-values-for-identical-strings/50536111#50536111

I think it makes sense attach the hasher object to the SRI class, and construct a new SRI object for every new line.

Reply:
Yep, `hashlib` objects are not reentrant. I just don't think that SRI should be a hashing object in of itself. The inverse is that a `hashlib` object does not implement checksum validation: and nor should it.

Possibly providing a `SRI.check` function adaptor around a readable may make sense?

If you make `SRI` have the shape of a `Hash` protocol, you can use a generic implementation of a `DigestError`.

```py
class HasBinaryRead(Protocol):
    def read(self, __size: int = -1) -> bytes | None: ...

class Hash(Protocol):
    def digest(self) -> bytes: ...
    def hexdigest(self) -> str: ...

class DigestError(ValueError):
    def __init__(self, *, expected: Hash, actual: Hash, message: str = 'Mismatched digest'):
        self.__expected = expected
        self.__actual = actual
        self.__message = message

    @property
    def expected(self) -> Hash:
        return self.__expected

    @property
    def actual(self) -> Hash:
        return self.__actual

    @property
    def message(self) -> str:
        return self.__message

    def __str__(self) -> str:
        return f'''{self.__message}:
 - expected: {self.expected.hexdigest()}
 - actual  : {self.actual.hexdigest()}'''

class SRI:
    def check(self, io: HasBinaryRead, chunk: int = 4096) -> Iterator[bytes]:
        hasher = hashlib.new(self.algorithm)
        while buffer := io.read(chunk):
            hasher.update(buffer)
            yield buffer
        if hasher.digest() != self.digest():
            raise DigestError(
                actual = hasher,
                expected = sri,
                message = "SRI mismatch `{self}`",
            )

def download(uri: URI, sri: SRI, filename: Path) -> None:
    filename.parent.mkdir(parents=True, exist_ok=True)
    with urlopen(str(uri)) as ws:
        with filename.open('wb') as fp:
            for buffer in sri.check(fp):
                fp.write(buf)
    chmod(filename, stat.S_IRUSR | stat.S_IXUSR)
```

Reply:
Okay, I implemented a modified version of your suggestions.